false-positive: TS2339 on a class method missing from a union of a type and its own alias (`ZodType[any,any,any] | ZodTypeAny`, zod types.ts:1995,2009)

## Witness (currently blocking the zod row, `main` @ `8189f4f`)

`colinhacks/zod` at `93b0b689`, `include: ["src/types.ts"]`, `strict`, `target es2020`, `module esnext`, `moduleResolution bundler`:

```
src/types.ts(1995,18): error TS2339: Property '_parse' does not exist on type 'ZodType[any, any, any] | ZodTypeAny'.
src/types.ts(2009,37): error TS2339: Property '_parseSync' does not exist on type 'ZodType[any, any, any] | ZodTypeAny'.
```

`tsc@7.0.2` on the identical `tsconfig` is clean (exit 0, no diagnostics). Confirmed with the vendored oracle (`npm i typescript@7.0.2`, `node .../lib/tsc.js -p tsconfig.probe.json --noEmit --strict --pretty false`).

**`ZodTypeAny` is a type alias for exactly `ZodType[any, any, any]`** (`src/types.ts:44`, `export type ZodTypeAny = ZodType[any, any, any]`). So the reported union is a type unioned with an alias for itself — the two members are semantically identical, and `_parse`/`_parseSync` are real members of `ZodType` (the abstract `_parse` at `types.ts:131`, the concrete `_parseSync` at `types.ts:143` which calls `this._parse()`). tsz is failing property lookup against one representation of the union that it accepts against the other.

## Where it comes from

`ZodUnion[T]._parse` (`types.ts:1957-1965`):

```ts
export class ZodUnion[T extends ZodUnionOptions] extends ZodType[
  T[number]["_output"],
  ZodUnionDef[T],
  T[number]["_input"]
] {
  _parse(...): ... {
    const options = this._def.options;
    ...
    options.map((option, index) => option._parse(...))   // line 1995
    ...
    const parsedOption = option._parseSync(...)            // line 2009, sync branch
```

`ZodUnionOptions = [ZodTypeAny, ...ZodTypeAny[]]` (`types.ts:1937`), `ZodUnionDef[T extends ZodUnionOptions = [ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]]] { options: T }` (`types.ts:1938-1943`). `options` is `this._def.options`, typed `T` (a generic tuple type param bound by `ZodUnionOptions`), and `.map()`'s callback parameter is each element of that tuple/array — which tsz apparently types as the **union** `ZodType[any,any,any] | ZodTypeAny` (the tuple's fixed first slot and its rest-array slot, printed with two different display forms of the identical type) instead of collapsing both to one type before printing and before property lookup. `_parse`/`_parseSync` exist on both members structurally, so this looks like a lookup gate treating one of the two union member *representations* (raw `Application[Lazy[ZodType], [any,any,any]]` vs the `Lazy` alias reference `ZodTypeAny`) as not-yet-materialized and skipping it, rather than a genuine relation failure — the same "no materialize-or-defer (apparent-type) gateway" pattern tracked in #15396.

## What did NOT reproduce (three reduction attempts, in order of added complexity)

All three are clean on tsz and (trivially) on tsc, so the trigger needs more of the real shape than any of these:

1. Two-arg class + tuple-rest `.map()`, no defaults, no abstract member:
   ```ts
   class ZodType[A, B, C] { _parse(): number { return 1; } }
   type ZodTypeAny = ZodType[any, any, any];
   type Options = [ZodTypeAny, ...ZodTypeAny[]];
   function useOptions[T extends Options](options: T) {
     options.map((option) => option._parse());
   }
   ```
2. Same, but with the real `_def: Def` field + `ZodUnionDef`-shaped interface with a longer-tuple default that doesn't match its own constraint's arity (`[ZodTypeAny, ...ZodTypeAny[]]` constraint vs `[ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]]` default) and a class extending `ZodType[any, ZodUnionDef[T], any]`. Still clean.
3. Same as (2) but `_parse` made `abstract` on `ZodType` (matching the real file exactly) with a concrete `_parseSync` calling `this._parse()`, and the heritage args changed to `T[number]` in the `Output`/`Input` slots (closer to the real `T[number]["_output"]`/`T[number]["_input"]` indexed-access form). Still clean.

Reduction (3) is the closest structural match I found and it does not reproduce, so the missing ingredient is still unidentified — possibly the real `T[number]["_output"]`/`["_input"]` **indexed-access** (vs. bare `T[number]`), the real multi-hundred-line `ZodType` base class surface, or some cross-reference elsewhere in the 3581-line file that this issue's witness depends on. Do not restart from a clean-room repro without first trying to reduce the real file directly (delete-and-recheck, one member/statement at a time, ~config).

## Also noteworthy: #16044 (the issue this witness superseded) is now fixed

The two `TS2322` diagnostics #16044 tracked (`types.ts:230,260`, `this._refinement(...)` losing `ZodEffects`'s type arguments) are **gone on current `main` (`8189f4f`)** — re-measured fresh with the identical fixture/config #16044 used. Something that landed between #16044's original measurement (`470022f`) and now (`8189f4f`) already fixed it, most likely #16047's intersection-inference fix given the "positional vs `inferFromTypes`" shape overlap, but I did not bisect to confirm which PR. Closing #16044 with this evidence; if a future session wants the exact fixing commit, `git bisect` between `470022f` and `8189f4f` on the standalone `types.ts` fixture would nail it in a handful of steps.

## Reproduction, cold-startable

```bash
git clone --filter=blob:none https://github.com/colinhacks/zod.git && cd zod && git checkout 93b0b689
cat > tsconfig.probe.json <<'EOF'
{"compilerOptions":{"strict":true,"target":"es2020","module":"esnext",
 "moduleResolution":"bundler","noEmit":true,"skipLibCheck":true},
 "include":["src/types.ts"]}
EOF
tsz --noEmit -p tsconfig.probe.json          # 2 x TS2339 at 1995,2009
npx typescript@7.0.2 tsc -p tsconfig.probe.json   # clean
```

## Provenance

Machine: cloud
Assistant: claude-code
Model: claude-sonnet-5
Effort: medium
AgentName: ClaudeT

--- a/crates/tsz-solver/src/evaluation/evaluate/application.rs
+++ b/crates/tsz-solver/src/evaluation/evaluate/application.rs
@@ -207,6 +207,35 @@ impl<'a, R: TypeResolver> TypeEvaluator<'a, R> {
                 self.apparent_conditional_branch = saved_apparent;
                 self.decrement_def_depth(def_id);
 
+                // A class application that evaluated to a structural object which
+                // dropped the class's nominal symbol is a degraded, partially
+                // built instance body: the class's instance type carried only its
+                // annotated fields (no methods, no nominal identity) when this
+                // application forced its resolution — a circular import can make a
+                // sibling's build reach `Class<Args>` before the class publishes
+                // its final type. Surfacing that partial object (e.g. as a
+                // `[X, ...X[]][number]` union member beside the complete
+                // representation) yields spurious `TS2339`s for every missing
+                // method (issue #16055). Discard it: purge whatever the body
+                // evaluation cached under `(def, args)` so a later evaluation
+                // recomputes against the finished body, taint the run so nothing
+                // persists this partial, and keep the application opaque so a
+                // property access re-resolves it. Mirrors the Phase-4.5 in-flight
+                // sentinel's `mark_unresolved_def_seen()`/return-opaque shape.
+                if result != original_type_id
+                    && matches!(
+                        self.resolver.get_def_kind(def_id),
+                        Some(crate::def::DefKind::Class)
+                    )
+                    && self.application_result_dropped_nominal_symbol(result)
+                {
+                    if let Some(db) = self.query_db {
+                        db.invalidate_application_eval_cache_for_def(def_id);
+                    }
+                    self.mark_unresolved_def_seen();
+                    return original_type_id;
+                }
+
                 // Phase 7 — display-alias bookkeeping. Skip entirely when
                 // the result is the original `Application` itself (the
                 // historical `if result != original_type_id` gate).
@@ -238,6 +267,21 @@ impl<'a, R: TypeResolver> TypeEvaluator<'a, R> {
                 }
                 result
             }
+        }
+    }
+
+    /// True when `result` is a structural object (`Object`/`ObjectWithIndex`)
+    /// carrying no nominal `symbol`. For a class application this is the
+    /// signature of an instance body instantiated from a partial,
+    /// mid-construction source: the complete instance keeps its class
+    /// `SymbolId`, whereas the annotated-fields-only snapshot produced before
+    /// the class finishes building loses it (issue #16055).
+    fn application_result_dropped_nominal_symbol(&self, result: TypeId) -> bool {
+        match self.interner.lookup(result) {
+            Some(TypeData::Object(shape_id)) | Some(TypeData::ObjectWithIndex(shape_id)) => {
+                self.interner.object_shape(shape_id).symbol.is_none()
+            }
+            _ => false,
         }
     }
 

